src/stan/io/json/json_data_handler.hpp

```diff
@@ -7,174 +7,527 @@
+    std::vector<std::string> stack(keys);
+    std::string key;
+    stack.pop_back();
+    while (!stack.empty()) {
+      key = boost::algorithm::join(stack, ".");
+      if (slot_types_map[key] == meta_type::ARRAY_OF_TUPLES)
+        return true;
+      stack.pop_back();
+    }
+    return false;
+  }
+
+  array_dims get_outer_dims(const std::vector<std::string>& keys) {
+    std::vector<std::string> stack(keys);
+    std::string key;
+    stack.pop_back();
+    while (!stack.empty()) {
+      key = boost::algorithm::join(stack, ".");
+      if (slot_dims_map.count(key) == 1)
+        return slot_dims_map[key];
+      stack.pop_back();
+    }
+    key = boost::algorithm::join(keys, ".");
+    if (slot_dims_map.count(key) != 1)
+      unexpected_error(key);
+    return slot_dims_map[key];
+  }
+
+  void set_outer_dims(array_dims update) {
+    std::vector<std::string> stack = key_stack;
+    std::string key;
+    stack.pop_back();
+    while (!stack.empty()) {
+      key = boost::algorithm::join(stack, ".");
+      if (slot_dims_map.count(key) == 1)
+        break;
+      stack.pop_back();
+    }
+    if (stack.empty()) {
+      key = boost::algorithm::join(key_stack, ".");
+      unexpected_error(key);
+    }
+    slot_dims_map[key] = update;
+  }
+
+  void promote_to_double() {
+    if (int_slots_map[key_str()]) {
+      for (std::vector<int>::iterator it = values_i.begin();
+           it != values_i.end(); ++it)
+        values_r.push_back(*it);
+      array_start_r = array_start_i;
+    }
+    int_slots_map[key_str()] = false;
+  }
+
+  /* Save non-tuple vars and innermost tuple slots to vars_i and vars_r.
+   * For arrays of tuples we need to check that new elements are consistent
+   * with previous tuple elements.
+   */
+  void save_key_value_pair() {
+    if (0 == key_stack.size())
+      return;
+    if (slot_types_map.count(key_str()) < 1)
+      unexpected_error(key_str());
+    if (slot_types_map[key_str()] == meta_type::TUPLE
+        || slot_types_map[key_str()] == meta_type::ARRAY_OF_TUPLES) {
+      {
+      }
+    } else {
+      bool is_int = int_slots_map[key_str()];
+      bool is_new
+          = (vars_r.count(key_str()) == 0 && vars_i.count(key_str()) == 0);
+      bool is_real = vars_r.count(key_str()) == 1;
+      bool was_int = vars_i.count(key_str()) == 1;
+      std::vector<size_t> dims;
+      if (slot_dims_map.count(key_str()) == 1)
+        dims = slot_dims_map[key_str()].dims;
+      if (is_new) {
+        var_types_map[key_str()] = slot_types_map[key_str()];
+        if (is_int) {
+          std::pair<std::vector<int>, std::vector<size_t>> pair;
+          pair = make_pair(values_i, dims);
+          vars_i[key_str()] = pair;
+        } else {
+          std::pair<std::vector<double>, std::vector<size_t>> pair;
+          pair = make_pair(values_r, dims);
+          vars_r[key_str()] = pair;
+        }
+      } else {
+        if (!is_array_tuples(key_stack)) {
```

Comment:
This seems suspect to me - why does the `redefine variable` error only occur for non-array-tuples?

Reply:
I think logic at line 222 is the problem